### DISCUSSION TITLE
Draggable list demo expansion

### DISCUSSION THREAD

**Discussion by g-danna** [pmndrs/react-spring#1145] [Support]
Hello react-spring community! Let me start by thanking the maintainers for their hard work on this magic library!

I've been working on and expanding the [draggable list demo](https://codesandbox.io/embed/r5qmj8m6lq) with a few features. Some I've been (to some extent) successful at, others... eh, not really. I'll use this post as a way to **share examples** for anyone who might find them useful, and, at the same time, to **ask for some help** in solving some issues.

Features added on top of the [`useSprings`](https://www.react-spring.io/docs/hooks/use-springs) [demo](https://codesandbox.io/embed/r5qmj8m6lq):

- [x]  Typescript version
- [x] From absolutely to relatively positioned items
- [ ]  **Toggle sort mode**
    - [ ] When set to `true`, shrinks the list from the content `auto` height to a fixed one (almost done)
    - [ ] `onSortComplete(newOrder: number[])` callback with the new order that can be used to re-render the list with the new order (almost done)
- [ ]  Scroll window up/down automatically when approaching the top/bottom respectively

[**CodeSandbox Link**](https://codesandbox.io/s/react-spring-draggable-list-typescript-accordion-h1pdt)

![b5dba6f46057c3ae2c77a10a1acdad2f](https://user-images.githubusercontent.com/23014664/92996083-f6f4c000-f508-11ea-9d24-9e5dae14a30c.gif)

## Toggle sort mode

Say you want to sort long paragraphs of selectable text, or anything that has `auto` height or a very long height, then I think it's useful to be able to toggle a 'sort mode' in which the list items change props and have a fixed height.

For example, on LinkedIn, if you have an education entry with a very long description and your browser window is smaller than its height, you won't be able to sort it at all. This is also where the relatively positioned elements and the automatic edge scrolling features come into play.

**Issues to solve**:

1. If the user clicks the toggle sort mode button fast enough, the **height animation will stop mid-way**. The items are now sortable but have don't have the wanted height. I used the async [spring scripting](https://www.react-spring.io/docs/hooks/use-spring#async-chainsscripts) for this, and I don't know if there are some gotchas I don't know, or if it's a bug.
2. When exiting of sort-mode, only if the the `onSortComplete(newOrder)` callback is provided, the list elements change into the new order **either animating** (ignoring the `immediate` attribute) or in a weird two-frame animation, which you can see below. Here, there definitely is some gotcha I don't understand about react's lifecycles, and would appreciate any tip. I've read [the complete guide to `useEffect()`](http://www.cristo.top/a-complete-guide-to-useeffect/) but I can't wrap my head around it completely apparently.

![a04f18dc5d6717a3454f63b733d990aa](https://user-images.githubusercontent.com/23014664/92996085-fe1bce00-f508-11ea-9c46-49f522b38f03.gif)

Thank you in advance for your help! ✌️

**Comment by aleclarson** [comment]
Interesting demo! I'll be sure to test this out on the next RC before it's released to see if it surfaces any bugs. 👍

**Reply by YashasKirito** [reply]
@guglielmodanna Were you able to get the scroll window feature to work? I've been trying to create that animation but i haven't gotten any great ideas as to how to approach this problem.

Any suggestions would help. Thanks! ♥️
